### PR TITLE
Convert Yieldspace pricing model to fixed point

### PR DESCRIPTION
**Pull request by sentilesdal**
- convert yieldspace
- add some constants
- add more fixed point dataclasses
- add error
- add fp base pricing model
- add new tets and fix things

### CODE REVIEW

File: INSTALL.md [hunk 1]
```diff
@@ -19,6 +19,7 @@ source .venv/bin/activate
 Once you're in your favored virtual environment, install the project dependencies:
 
 ```bash
+python -m pip install --upgrade pip
```
Comment (wakamex) [inline INSTALL.md:22]: I wanted to leave this out, cus some packages still don't fully support the new install conventions. so why push for something that isn't even the default yet, when we're pinned to a python (major) version that came out Oct. 5, 2020?

File: INSTALL.md [hunk 2]
```diff
@@ -44,7 +45,7 @@ docker run -it --name elf-simulations-dev --rm --volume $(pwd):/app/ --net=host
 ## Apeworks and Contract Integration
 
 NOTE: The Hyperdrive solidity implementation is currently under security review, and thus is not available publicly.
-The following instructions will not work for anyone who is not a member of Element Finance.
+The following instructions will not work for anyone who is not a member of Delv.
```
Comment (wakamex) [inline INSTALL.md:48]: I'd prefer these to be their own PR, but I'll accept this for now

File: elfpy/markets/hyperdrive/hyperdrive_market.py [hunk 3]
```diff
@@ -139,30 +163,165 @@ def __setitem__(self, key, value):
...
         self.withdraw_capital += delta.withdraw_capital
         self.withdraw_interest += delta.withdraw_interest
-
+        # checkpointing
         for mint_time, delta_checkpoint in delta.long_checkpoints.items():
             self.checkpoints[mint_time].long_base_volume += delta_checkpoint
         for mint_time, delta_checkpoint in delta.short_checkpoints.items():
             self.checkpoints[mint_time].short_base_volume += delta_checkpoint
+        for mint_time, delta_supply in delta.total_supply_longs.items():
+            self.total_supply_longs[mint_time] += delta_supply
```
Comment (wakamex) [inline elfpy/markets/hyperdrive/hyperdrive_market.py:191]: why are these added here? seems unrelated to FixedPoint

Comment (wakamex) [reply elfpy/markets/hyperdrive/hyperdrive_market.py:191]: oh, the diffs just got cut up, nevermind

File: elfpy/utils/math/fixed_point.py [hunk 4]
```diff
@@ -4,6 +4,8 @@
 import copy
 from typing import Union
 
+import elfpy.errors.errors as errors
```
Comment (wakamex) [inline elfpy/utils/math/fixed_point.py:7]: based

File: elfpy/utils/math/fixed_point.py [hunk 5]
```diff
@@ -130,44 +130,97 @@ def __rpow__(self, other: int | FixedPoint) -> FixedPoint:
...
+    EXP_MIN = -42139678854452767622  # floor(log(0.5e-18)*1e18)
     ONE_18 = 1 * 10**18
 
     @staticmethod
     def add(a: int, b: int) -> int:
         """Add two fixed-point numbers in 1e18 format."""
         c = a + b
-        if c < a or a > FixedPointMath.INT_MAX - b:
+        # solidity has this: `if c < a or a > FixedPointMath.INT_MAX - b`
+        # however, we allow negavitve values sometimes, which trips up that check.  python 3 also
```
Comment (wakamex) [inline elfpy/utils/math/fixed_point.py:210]: typo

**Review by wakamex — APPROVED** [review]
lgtm

**Review by dpaiton — APPROVED** [review]

File: INSTALL.md (hunk 2 above)
Comment (dpaiton) [reply INSTALL.md:48]: har har

File: INSTALL.md (hunk 1 above)
Comment (dpaiton) [reply INSTALL.md:22]: Which packages don't like upgrading pip? Keep in mind that this is the latest version of pip _given_ the python version. I believe it is mostly updating package version collisions & dependency lists.
